File: android/expoview/src/main/java/host/exp/exponent/notifications/helpers/ExpoCronParser.java
package host.exp.exponent.notifications.helpers;

import com.cronutils.builder.CronBuilder;
import com.cronutils.model.Cron;

import java.util.HashMap;

import static com.cronutils.model.field.expression.FieldExpressionFactory.always;
import static com.cronutils.model.field.expression.FieldExpressionFactory.on;
import static com.cronutils.model.field.expression.FieldExpressionFactory.questionMark;

public class ExpoCronParser {
  public static Cron createCronInstance(HashMap<String, Object> options) {
    CronBuilder cronBuilder = CronBuilder.cron(ExpoCronDefinitionBuilder.getCronDefinition());

    if (options.get("year") instanceof Number) {
      cronBuilder.withYear(on(((Number) options.get("year")).intValue()));
    } else {
      cronBuilder.withYear(always());
    }

    if (options.get("hour") instanceof Number) {
      cronBuilder.withHour(on(((Number) options.get("hour")).intValue()));
    } else {
      cronBuilder.withHour(always());
    }

    if (options.get("minute") instanceof Number) {
      cronBuilder.withMinute(on(((Number) options.get("minute")).intValue()));
    } else {
      cronBuilder.withMinute(always());
    }

    if (options.get("second") instanceof Number) {
      cronBuilder.withSecond(on(((Number) options.get("second")).intValue()));
    } else {
      cronBuilder.withSecond(always());
    }

    if (options.get("month") instanceof Number) {
      cronBuilder.withMonth(on(((Number) options.get("month")).intValue()));
    } else {
      cronBuilder.withMonth(always());
    }

    if (options.get("day") instanceof Number) {
      cronBuilder.withDoM(on(((Number) options.get("day")).intValue()));
    } else if(options.containsKey("weekDay")) {
      cronBuilder.withDoM(questionMark());
    } else {
      cronBuilder.withDoM(always());
    }

    if (options.get("weekDay") instanceof Number) {
      cronBuilder.withDoW(on(((Number) options.get("weekDay")).intValue()));
    } else {
      cronBuilder.withDoW(questionMark());
    }

    return cronBuilder.instance();
  }
}

File: android/expoview/src/main/java/host/exp/exponent/notifications/helpers/ExpoCronParser.kt
package host.exp.exponent.notifications.helpers

import com.cronutils.builder.CronBuilder
import com.cronutils.model.Cron
import com.cronutils.model.field.expression.FieldExpressionFactory
import java.util.*

object ExpoCronParser {
  @JvmStatic fun createCronInstance(options: HashMap<String?, Any?>): Cron {
    val cronBuilder = CronBuilder.cron(ExpoCronDefinitionBuilder.cronDefinition)

    val year = options["year"]
    if (year is Number) {
      cronBuilder.withYear(FieldExpressionFactory.on(year.toInt()))
    } else {
      cronBuilder.withYear(FieldExpressionFactory.always())
    }

    val hour = options["hour"]
    if (hour is Number) {
      cronBuilder.withHour(FieldExpressionFactory.on(hour.toInt()))
    } else {
      cronBuilder.withHour(FieldExpressionFactory.always())
    }

    val minute = options["minute"]
    if (minute is Number) {
      cronBuilder.withMinute(FieldExpressionFactory.on(minute.toInt()))
    } else {
      cronBuilder.withMinute(FieldExpressionFactory.always())
    }

    val second = options["second"]
    if (second is Number) {
      cronBuilder.withSecond(FieldExpressionFactory.on(second.toInt()))
    } else {
      cronBuilder.withSecond(FieldExpressionFactory.always())
    }

    val month = options["month"]
    if (month is Number) {
      cronBuilder.withMonth(FieldExpressionFactory.on(month.toInt()))
    } else {
      cronBuilder.withMonth(FieldExpressionFactory.always())
    }

    val day = options["day"]
    if (day is Number) {
      cronBuilder.withDoM(FieldExpressionFactory.on(day.toInt()))
    } else if (options.containsKey("weekDay")) {
      cronBuilder.withDoM(FieldExpressionFactory.questionMark())
    } else {
      cronBuilder.withDoM(FieldExpressionFactory.always())
    }

    val weekDay = options["weekDay"]
    if (weekDay is Number) {
      cronBuilder.withDoW(FieldExpressionFactory.on(weekDay.toInt()))
    } else {
      cronBuilder.withDoW(FieldExpressionFactory.questionMark())
    }

    return cronBuilder.instance()
  }
}
